Redesign of the `docs` as an `mdBook`

haixuanTao opened the pull request:
This PR moves many of the documentation into a `mdBook` format version for ease of readibility.

It includes a github action workflows that publish the book on the default `gh-pages` branch to be publish on github page.

phil-opp commented on docs/src/communication-layer.md:
```diff
@@ -0,0 +1,30 @@
+# Middleware (communication) layer abstraction (MLA)
+
```
Maybe we should link to https://github.com/dora-rs/dora/discussions/53 here?

haixuanTao commented:
So, I have linked many example to example that are in the coordinator folder. I will wait for the C Node API to be merged so that I can link to dataflow that are in the example folder. instead of the coordinator folder.

phil-opp reviewed (approved):
Looks great!

haixuanTao commented:
Ok! I have
- rebased on the latest main branch
- Linked the latest c-dataflow 
- Added documentation for the c node API.

I think we can merge this one.

I will follow up with a PR adding Python dataflow and Rust dataflow as well as linking them in the documentation.

phil-opp commented:
Great, thanks! I'll do a quick review.



> I will follow up with a PR adding Python dataflow and Rust dataflow as well as linking them in the documentation.

I'm currently working on a Rust dataflow example already.

phil-opp commented on docs/src/c-api.md:
```diff
@@ -0,0 +1,96 @@
+    void *output_context,
+    const void *operator_context)
+{
+  ...
```
```suggestion
  // handle the input ...
  // (sending outputs is possible using `output_fn_raw`)
  // (the `operator_context` is the pointer created in `dora_init_operator`, i.e., a counter in our case)
```

phil-opp commented on docs/src/installation.md:
````diff
@@ -0,0 +1,27 @@
+cargo build -p dora-runtime --release
+```
+
+### 3. Add those binaries to your path
+
+```bash
+export PATH=$PATH:$(pwd)/target/release
+```
````
```suggestion
### 3. Add those binaries to your path

This step is optional. You can also refer to the executables using their full path or copy them somewhere else.

```bash
export PATH=$PATH:$(pwd)/target/release
```
```

phil-opp commented on docs/src/python-api.md:
```diff
@@ -0,0 +1,85 @@
+# Python API
+
+## Cutom Node
```
```suggestion
## Custom Node
```

phil-opp commented on docs/src/python-api.md:
```diff
@@ -0,0 +1,85 @@
+
+#### `.next()` or `__next__()` as an iterator
+
+`.next()` gives you the next input that the node has received.
```
```suggestion
`.next()` gives you the next input that the node has received. It blocks until the next input becomes available.
```

haixuanTao replied:
So for python, it will not block. It will send a `None` value if there is no available input in the `tokio::channel`. I will add that to the documentation

phil-opp replied:
Are you sure? I think we're doing `blocking_recv` on the channel, which will only return `None` when the channel is closed.

haixuanTao replied:
Yeah, you're right. I just checked the documentation again, and it in fact, block it until the sender is dropped, this is where it sends None.

phil-opp commented on docs/src/python-api.md:
````diff
@@ -0,0 +1,85 @@
+{{#include ../../binaries/coordinator/examples/graphs/python_test.yml:12:17}}
+```
+
+## Operator
+
+The operator API gives you a framework for operator that is going to be managed by `dora`. This framework enable us to make optimisation and provide advanced features.
+
+An operator requires an `on_input` method and requires to return a `DoraStatus` of 0 or 1, depending of it needs to continue or stop.
+
+```python
+class Operator:
+    def on_input(
+        self,
+        input_id: str,
+        value: bytes,
+        send_output: Callable[[str, bytes], None],
+    ) -> DoraStatus:
+```
+
+> For Python, we recommend to allocate the operator on a single runtime. A runtime will share the same GIL with several operators making those operators run almost sequentially. See: [https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#deadlocks](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#deadlocks)
+### Try it out!
+
+- Create an operator python file called `op.py`:
+```python
+{{#include ../../examples/python-operator/op.py}}
+```
+
+- Link it in your graph as:
+```yaml
+{{#include ../../binaries/coordinator/examples/graphs/mini-dataflow.yml:67:73}}
+```
````
We want users to use operators instead of custom nodes if possible, so I think we should describe them first.

phil-opp commented on docs/src/rust-api.md:
````diff
@@ -0,0 +1,117 @@
+        - time
+```
+
+## Operator 
````
This section should be first since we want users to use operators by default.

phil-opp reviewed (approved):
Looks good! I left some minor suggestions.

haixuanTao commented on docs/src/c-api.md:
```diff
@@ -0,0 +1,96 @@
+{{#include ../../examples/c-dataflow/README.md:26:35}}
+
+
+## Operator
```
ok sure!

haixuanTao commented on docs/src/c-api.md:
````diff
@@ -0,0 +1,96 @@
+char *data;
+size_t data_len;
+read_dora_input_data(input, &data, &data_len);
+```
````
Thanks! I actually miss read that part
